### DISCUSSION TITLE
Custom Material Import with Arbitrary Permittivity and Conductivity Distribution

### DISCUSSION THREAD

**Discussion by cyasing** [NanoComp/meep#2310] [Q&A] [answered]
I want to import data of a complex material (with arbitrary permittivity and conductivity) defined over a uniformly distributed set of grid points. The two data files have permittivity / conductivity values w.r.t 2D coordinates. Coordinates are mentioned since the material does not extend beyond these coordinates (essentially geometry and material data in one). The only two options that came close were:

1. material_function: Does not seem viable, since the values are arbitrary.
2. epsilon_input_file: Only applicable for region without geometry. Also, no option for conductivity file. 

What should I do?

**Comment by smartalecH** [comment, marked as answer]
I would use a `material_function` and just wrap your discrete samples with some arbitrary interpolator (eg linear interpolation is what meep uses everywhere).

Remember, meep's philosophy is to promote an _illusion of continuity_. So it expects continuous representations as inputs, even if you have to fake it with interpolation (as that's what meep does under the hood using _pervasive interpolation_)

**Reply by cyasing** [reply]
Yeah, that did seem like the only way out.

**Reply by cyasing** [reply]
@smartalecH I tried getting this to work, but the problem is that my data is too large (~16,000 entries). This is what I am doing:

```
def custom_material(p):
    BR_con_pd = pd.read_excel('conductivity.xlsx') #Data has columns for x, y, conductivity values
    BR_per_pd = pd.read_excel('permittivity.xlsx') #Data has columns for x, y, permittivity values
    bcon = np.array(BR_con_pd)
    bper = np.array(BR_per_pd)

    if round(p.x, 2) in bcon[:,0].round(2) and round(p.y, 2) in bcon[:,1].round(2):
      #Rounded because MEEP and my data have different precision and step size between entries
        con_val_here = bcon[idx,2]
        per_val_here = bper[idx,2]
        return mp.Medium(epsilon=per_val_here, D_conductivity=2*np.pi*freq*con_val_here/per_val_here)
    else:
     #Interpolation code here
        return mp.Medium(epsilon=per_val_here, D_conductivity=2*np.pi*freq*con_val_here/per_val_here)
```
Please tell me what I can do differently to make it faster.

**Reply by smartalecH** [reply]
What do you mean, "too large"? As in, it takes too long to build the geometry? Why not just downsample your original data?

**Reply by cyasing** [reply]
Yes, the geometry does take too long to build. I cannot down-sample it too much. The 'custom material' is actually a brain's dielectric profile.

**Reply by tulchhi-ram** [reply]
Did you found a way to improve this. I am simulating a media with space dependent media. When i insert as a slab with each slab a grid point, 
when I simulate this its working but too may objects taking very high time to geometry set. So pass the media as function its taking the epsilon value but not complex values. 

"The code in which it ignoring the conductivity term" 
def epsi(p):
    i=int(p.x)
    
    if abs(p.x)>7000:
        ep=mp.Medium(index=1)
    else: 
        epsr=sigme[i+6999,0]
        epsim=sigme[i+6999,1]
        D_conductivity=epsim
        susceptibilities = [mp.LorentzianSusceptibility(
            frequency=fcen, gamma=0, sigma=D_conductivity)]
        #ep=epsr+1j*epsim
        ep=mp.Medium(epsilon=epsr, E_susceptibilities=susceptibilities)
    return ep

geometry=[mp.Block(size=mp.Vector3(14000,sy-0*dpml,0), center=mp.Vector3(0,0,0),material=epsi)]

sim = mp.Simulation(cell_size=cell, boundary_layers=pml_layers, sources=sources,
                    resolution=resolution,geometry=geometry,symmetries=symmetries,Courant=0.2)"


"here its working"
geometry = []
sigm=np.zeros((14000))
for i in range(14000):

    D_conductivity=2*np.pi*fcen*sigme[i,1]/sigme[i,0]
    sigm[i]=D_conductivity
    susceptibilities = [mp.LorentzianSusceptibility(
        frequency=fcen, gamma=0, sigma=D_conductivity)]

    geometry.append(mp.Block(size=mp.Vector3(1,sy-0*dpml,0), center=mp.Vector3(-7000+i,0,0),
                                material=mp.Medium(epsilon=sigme[i,0],E_susceptibilities=susceptibilities)))



#sim = mp.Simulation(cell_size=cell, boundary_layers=pml_layers, sources=sources,
#                    resolution=resolution, geometry=geometry,symmetries=symmetries,Courant=0.1)

sim = mp.Simulation(cell_size=cell, boundary_layers=pml_layers, sources=sources,
                    resolution=resolution,geometry=geometry,symmetries=symmetries,Courant=0.1)


Also i have tried to pass it to material function but not working

**Comment by recapture2018** [comment]
这是来自QQ邮箱的假期自动回复邮件。
 
您好，我最近正在休假中，无法亲自回复您的邮件。我将在假期结束后，尽快给您回复。
